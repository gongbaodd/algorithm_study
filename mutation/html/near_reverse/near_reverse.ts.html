<!DOCTYPE html>
<html><head><title>near_reverse/near_reverse.ts
 - Stryker report</title>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link rel="stylesheet" href="../strykerResources/bootstrap/css/bootstrap.min.css">
<link rel="stylesheet" href="../strykerResources/highlightjs/styles/default.css">
<link rel="stylesheet" href="../strykerResources/stryker/stryker.css"></head>
<body><img class="stryker-image" alt="Stryker" src="../strykerResources/stryker/stryker-80x80.png">
<div class="container-fluid"><div class="row"><div class="col-lg-12"><h1 class="display-4">near_reverse/near_reverse.ts
 
<small class="text-muted">- Stryker report</small></h1></div></div>
<ol class="breadcrumb"><li class="breadcrumb-item"><a href="../index.html">All files</a></li>
<li class="breadcrumb-item active">near_reverse/near_reverse.ts</li></ol>
<div class="row"><div class="col-lg-12"><div class="row"><div class="col-sm-11"><table class="table table-sm table-hover table-bordered table-no-top"><thead><tr><th style="width: 20%"><div><span>File / Directory</span></div></th>
<th colspan="2"><div><span>Mutation score</span></div></th>
<th class="rotate text-center" style="width: 50px"><div><span># Killed</span></div></th>
<th class="rotate text-center" style="width: 50px"><div><span># Survived</span></div></th>
<th class="rotate text-center" style="width: 50px"><div><span># Timeout</span></div></th>
<th class="rotate text-center" style="width: 50px"><div><span># No coverage</span></div></th>
<th class="rotate text-center" style="width: 50px"><div><span># Runtime errors</span></div></th>
<th class="rotate text-center" style="width: 50px"><div><span># Transpile errors</span></div></th>
<th class="rotate rotate-width-70 text-center" style="width: 70px"><div><span>Total detected</span></div></th>
<th class="rotate rotate-width-70 text-center" style="width: 70px"><div><span>Total undetected</span></div></th>
<th class="rotate rotate-width-70 text-center" style="width: 70px"><div><span>Total mutants</span></div></th></tr></thead>
<tbody><tr><td>near_reverse/near_reverse.ts</td>
<td><div class="progress"><div class="progress-bar bg-success" role="progressbar" aria-valuenow="100.00" aria-valuemin="0" aria-valuemax="100" style="width: 100.00%;">100.00
%</div></div></td>
<th class="text-center text-success" style="width: 10%">100.00</th>
<td class="text-center">8</td>
<td class="text-center">0</td>
<td class="text-center">5</td>
<td class="text-center">0</td>
<td class="text-center">8</td>
<td class="text-center">0</td>
<th class="text-center">13</th>
<th class="text-center">0</th>
<th class="text-center">21</th></tr>
</tbody></table></div></div>
<div class="row"><div class="col-lg-7"><div class="row legend"><form class="col-md-12" novalidate="novalidate">

<div class="form-check form-check-inline"><label class="form-check-label"><input class="form-check-input stryker-display" value="1" type="checkbox">
Killed
 
(8)</label></div>
<div class="form-check form-check-inline"><label class="form-check-label"><input class="form-check-input stryker-display" value="3" type="checkbox">
TimedOut
 
(5)</label></div>
<div class="form-check form-check-inline"><label class="form-check-label"><input class="form-check-input stryker-display" value="4" type="checkbox">
RuntimeError
 
(8)</label></div>

<a href="#" class="stryker-collapse-expand-all">Expand all</a></form></div>
<pre><code class="lang-javascript"><span class="bg-null">import { LNode } from &quot;../lib&quot;;

export function nearReverse(_head: LNode): LNode </span><a href="#" class="stryker-mutant-button" tabindex="0" title="Block" data-content="status: RuntimeError" data-mutant-status-annotation="secondary" data-mutant-status="4" data-mutant="0"><span class="badge badge-secondary">0</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="0">{}</span><span class="stryker-original-code" data-mutant="0"><span class="bg-null">{
  const head = _head;
  if (</span><a href="#" class="stryker-mutant-button" tabindex="0" title="IfStatement" data-content="status: RuntimeError" data-mutant-status-annotation="secondary" data-mutant-status="4" data-mutant="1"><span class="badge badge-secondary">1</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="1">true</span><a href="#" class="stryker-mutant-button" tabindex="0" title="IfStatement" data-content="status: RuntimeError" data-mutant-status-annotation="secondary" data-mutant-status="4" data-mutant="2"><span class="badge badge-secondary">2</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="2">false</span><a href="#" class="stryker-mutant-button" tabindex="0" title="PrefixUnaryExpression" data-content="status: RuntimeError" data-mutant-status-annotation="secondary" data-mutant-status="4" data-mutant="3"><span class="badge badge-secondary">3</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="3">head.next</span><span class="stryker-original-code" data-mutant="1"><span class="stryker-original-code" data-mutant="2"><span class="stryker-original-code" data-mutant="3"><span class="bg-null">!head.next</span></span></span></span><span class="bg-null">) </span><a href="#" class="stryker-mutant-button" tabindex="0" title="Block" data-content="status: RuntimeError" data-mutant-status-annotation="secondary" data-mutant-status="4" data-mutant="4"><span class="badge badge-secondary">4</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="4">{}</span><span class="stryker-original-code" data-mutant="4"><span class="bg-null">{
    return head;
  }</span></span><span class="bg-null">

  let pre = head;
  let current = head.next;
  let { next } = current;

  while (</span><a href="#" class="stryker-mutant-button" tabindex="0" title="WhileStatement" data-content="status: RuntimeError" data-mutant-status-annotation="secondary" data-mutant-status="4" data-mutant="5"><span class="badge badge-secondary">5</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="5">false</span><span class="stryker-original-code" data-mutant="5"><span class="bg-null">next</span></span><span class="bg-null">) </span><a href="#" class="stryker-mutant-button" tabindex="0" title="Block" data-content="status: TimedOut" data-mutant-status-annotation="warning" data-mutant-status="3" data-mutant="6"><span class="badge badge-warning">6</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="6">{}</span><span class="stryker-original-code" data-mutant="6"><span class="bg-warning-light">{
    const nNext = next.next;
    </span><a href="#" class="stryker-mutant-button" tabindex="0" title="ConditionalExpression" data-content="status: Killed" data-mutant-status-annotation="success" data-mutant-status="1" data-mutant="7"><span class="badge badge-success">7</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="7">false</span><a href="#" class="stryker-mutant-button" tabindex="0" title="ConditionalExpression" data-content="status: Killed" data-mutant-status-annotation="success" data-mutant-status="1" data-mutant="8"><span class="badge badge-success">8</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="8">true</span><span class="stryker-original-code" data-mutant="7"><span class="stryker-original-code" data-mutant="8"><span class="bg-warning-light">next.next = current</span></span></span><span class="bg-warning-light">;
    </span><a href="#" class="stryker-mutant-button" tabindex="0" title="ConditionalExpression" data-content="status: Killed" data-mutant-status-annotation="success" data-mutant-status="1" data-mutant="9"><span class="badge badge-success">9</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="9">false</span><a href="#" class="stryker-mutant-button" tabindex="0" title="ConditionalExpression" data-content="status: Killed" data-mutant-status-annotation="success" data-mutant-status="1" data-mutant="10"><span class="badge badge-success">10</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="10">true</span><span class="stryker-original-code" data-mutant="9"><span class="stryker-original-code" data-mutant="10"><span class="bg-warning-light">current.next = nNext</span></span></span><span class="bg-warning-light">;
    </span><a href="#" class="stryker-mutant-button" tabindex="0" title="ConditionalExpression" data-content="status: Killed" data-mutant-status-annotation="success" data-mutant-status="1" data-mutant="11"><span class="badge badge-success">11</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="11">false</span><a href="#" class="stryker-mutant-button" tabindex="0" title="ConditionalExpression" data-content="status: Killed" data-mutant-status-annotation="success" data-mutant-status="1" data-mutant="12"><span class="badge badge-success">12</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="12">true</span><span class="stryker-original-code" data-mutant="11"><span class="stryker-original-code" data-mutant="12"><span class="bg-warning-light">pre.next = next</span></span></span><span class="bg-warning-light">;

    </span><a href="#" class="stryker-mutant-button" tabindex="0" title="ConditionalExpression" data-content="status: Killed" data-mutant-status-annotation="success" data-mutant-status="1" data-mutant="13"><span class="badge badge-success">13</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="13">false</span><a href="#" class="stryker-mutant-button" tabindex="0" title="ConditionalExpression" data-content="status: Killed" data-mutant-status-annotation="success" data-mutant-status="1" data-mutant="14"><span class="badge badge-success">14</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="14">true</span><span class="stryker-original-code" data-mutant="13"><span class="stryker-original-code" data-mutant="14"><span class="bg-warning-light">pre = current</span></span></span><span class="bg-warning-light">;
    </span><a href="#" class="stryker-mutant-button" tabindex="0" title="ConditionalExpression" data-content="status: TimedOut" data-mutant-status-annotation="warning" data-mutant-status="3" data-mutant="15"><span class="badge badge-warning">15</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="15">false</span><a href="#" class="stryker-mutant-button" tabindex="0" title="ConditionalExpression" data-content="status: TimedOut" data-mutant-status-annotation="warning" data-mutant-status="3" data-mutant="16"><span class="badge badge-warning">16</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="16">true</span><span class="stryker-original-code" data-mutant="15"><span class="stryker-original-code" data-mutant="16"><span class="bg-warning-light">current = nNext as LNode</span></span></span><span class="bg-warning-light">;
    </span><a href="#" class="stryker-mutant-button" tabindex="0" title="ConditionalExpression" data-content="status: TimedOut" data-mutant-status-annotation="warning" data-mutant-status="3" data-mutant="17"><span class="badge badge-warning">17</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="17">false</span><a href="#" class="stryker-mutant-button" tabindex="0" title="ConditionalExpression" data-content="status: TimedOut" data-mutant-status-annotation="warning" data-mutant-status="3" data-mutant="18"><span class="badge badge-warning">18</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="18">true</span><span class="stryker-original-code" data-mutant="17"><span class="stryker-original-code" data-mutant="18"><span class="bg-warning-light">next = </span><a href="#" class="stryker-mutant-button" tabindex="0" title="ConditionalExpression" data-content="status: RuntimeError" data-mutant-status-annotation="secondary" data-mutant-status="4" data-mutant="19"><span class="badge badge-secondary">19</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="19">false</span><a href="#" class="stryker-mutant-button" tabindex="0" title="ConditionalExpression" data-content="status: RuntimeError" data-mutant-status-annotation="secondary" data-mutant-status="4" data-mutant="20"><span class="badge badge-secondary">20</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="20">true</span><span class="stryker-original-code" data-mutant="19"><span class="stryker-original-code" data-mutant="20"><span class="bg-warning-light">current?.next ?? null</span></span></span></span></span><span class="bg-warning-light">;
  }</span></span><span class="bg-null">

  return head;
}</span></span><span class="bg-null">
</code></pre></div>
<div class="col-lg-5"><table class="table table-sm table-hover mutant-table"><thead><tr><th>#</th>
<th>Mutator</th>
<th>State</th>
<th>Location</th>
<th>Original</th>
<th>Replacement</th></tr></thead>
<tbody><tr><th>0</th>
<td>Block</td>
<td>RuntimeError</td>
<td>2
:
49</td>
<td><code>{
   ...  ;
}</code></td>
<td><code>{}</code></td></tr>
<tr><th>1</th>
<td>IfStatement</td>
<td>RuntimeError</td>
<td>4
:
6</td>
<td><code>!    .</code></td>
<td><code></code></td></tr>
<tr><th>2</th>
<td>IfStatement</td>
<td>RuntimeError</td>
<td>4
:
6</td>
<td><code>!    .</code></td>
<td><code></code></td></tr>
<tr><th>3</th>
<td>PrefixUnaryExpression</td>
<td>RuntimeError</td>
<td>4
:
6</td>
<td><code>!    .</code></td>
<td><code>.</code></td></tr>
<tr><th>4</th>
<td>Block</td>
<td>RuntimeError</td>
<td>4
:
18</td>
<td><code>{
   ...;
  }</code></td>
<td><code>{}</code></td></tr>
<tr><th>5</th>
<td>WhileStatement</td>
<td>RuntimeError</td>
<td>12
:
9</td>
<td><code></code></td>
<td><code></code></td></tr>
<tr><th>6</th>
<td>Block</td>
<td>TimedOut</td>
<td>12
:
15</td>
<td><code>{
   ...;
  }</code></td>
<td><code>{}</code></td></tr>
<tr><th>7</th>
<td>ConditionalExpression</td>
<td>Killed</td>
<td>14
:
4</td>
<td><code>.     =</code></td>
<td><code></code></td></tr>
<tr><th>8</th>
<td>ConditionalExpression</td>
<td>Killed</td>
<td>14
:
4</td>
<td><code>.     =</code></td>
<td><code></code></td></tr>
<tr><th>9</th>
<td>ConditionalExpression</td>
<td>Killed</td>
<td>15
:
4</td>
<td><code>.     =</code></td>
<td><code></code></td></tr>
<tr><th>10</th>
<td>ConditionalExpression</td>
<td>Killed</td>
<td>15
:
4</td>
<td><code>.     =</code></td>
<td><code></code></td></tr>
<tr><th>11</th>
<td>ConditionalExpression</td>
<td>Killed</td>
<td>16
:
4</td>
<td><code>.     =</code></td>
<td><code></code></td></tr>
<tr><th>12</th>
<td>ConditionalExpression</td>
<td>Killed</td>
<td>16
:
4</td>
<td><code>.     =</code></td>
<td><code></code></td></tr>
<tr><th>13</th>
<td>ConditionalExpression</td>
<td>Killed</td>
<td>18
:
4</td>
<td><code>=</code></td>
<td><code></code></td></tr>
<tr><th>14</th>
<td>ConditionalExpression</td>
<td>Killed</td>
<td>18
:
4</td>
<td><code>=</code></td>
<td><code></code></td></tr>
<tr><th>15</th>
<td>ConditionalExpression</td>
<td>TimedOut</td>
<td>19
:
4</td>
<td><code>=</code></td>
<td><code></code></td></tr>
<tr><th>16</th>
<td>ConditionalExpression</td>
<td>TimedOut</td>
<td>19
:
4</td>
<td><code>=</code></td>
<td><code></code></td></tr>
<tr><th>17</th>
<td>ConditionalExpression</td>
<td>TimedOut</td>
<td>20
:
4</td>
<td><code>=    ...   ??</code></td>
<td><code></code></td></tr>
<tr><th>18</th>
<td>ConditionalExpression</td>
<td>TimedOut</td>
<td>20
:
4</td>
<td><code>=    ...   ??</code></td>
<td><code></code></td></tr>
<tr><th>19</th>
<td>ConditionalExpression</td>
<td>RuntimeError</td>
<td>20
:
11</td>
<td><code>?.     ??</code></td>
<td><code></code></td></tr>
<tr><th>20</th>
<td>ConditionalExpression</td>
<td>RuntimeError</td>
<td>20
:
11</td>
<td><code>?.     ??</code></td>
<td><code></code></td></tr></tbody></table></div></div></div></div>
<footer>Generated with stryker-html-reporter generator. Visit the 
<a href="https://stryker-mutator.io" target="_blank">Stryker website</a></footer></div>
<script src="../strykerResources/popper.js/dist/umd/popper.min.js" defer="defer"></script>
<script src="../strykerResources/jquery/dist/jquery.slim.min.js" defer="defer"></script>
<script src="../strykerResources/bootstrap/js/bootstrap.min.js" defer="defer"></script>
<script src="../strykerResources/stryker.js" defer="defer"></script></body></html>