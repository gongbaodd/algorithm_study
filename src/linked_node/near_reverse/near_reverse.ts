import { LNode } from "../lib";

export function nearReverse(_head: LNode): LNode {
  const head = _head;
  if (!head.next) {
    return head;
  }

  let pre = head;
  let current = head.next;
  let { next } = current;

  while (next) {
    const nNext = next.next;
    next.next = current;
    current.next = nNext;
    pre.next = next;

    pre = current;
    current = nNext as LNode;
    next = current?.next ?? null;
  }

  return head;
}
