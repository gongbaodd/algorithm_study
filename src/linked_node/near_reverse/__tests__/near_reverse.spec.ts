import { nearReverse } from "../near_reverse";
import { init } from "../../lib";

describe("near Reverse", () => {
  describe("empty chain", () => {
    it("should return the origin chain", () => {
      const head = init([]);
      const res = nearReverse(head);

      expect(res).toEqual(head);
      expect(res.next).toBeNull();
    });
  });

  describe("normal chain", () => {
    it("should return as expected", () => {
      const head = nearReverse(init([1, 2, 3, 4, 5, 6, 7]));
      expect(head).toMatchSnapshot();
    });
  });
});
