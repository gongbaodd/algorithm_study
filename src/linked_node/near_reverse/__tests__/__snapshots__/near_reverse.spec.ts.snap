// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`near Reverse normal chain should return as expected 1`] = `
LNode {
  "data": 0,
  "next": LNode {
    "data": 2,
    "next": LNode {
      "data": 1,
      "next": LNode {
        "data": 4,
        "next": LNode {
          "data": 3,
          "next": LNode {
            "data": 6,
            "next": LNode {
              "data": 5,
              "next": LNode {
                "data": 7,
                "next": null,
              },
            },
          },
        },
      },
    },
  },
}
`;
